### PR TITLE
[mlir][tensor] Enhance pack/unpack simplification for identity outer_dims_perm cases.

### PR DESCRIPTION
**Pull request by hanhanW**
They can be simplified to reshape ops if outer_dims_perm is an identity permutation. The revision adds a `isIdentityPermutation` method to IndexingUtils.

### CODE REVIEW

**Review by Max191 — APPROVED** [review]
Nice, thanks for adding this!

**Review by chelini — COMMENTED** [review]
I am not sure about this change. If outer dims perm is not a permutation in the first place, would it make more sense to drop it from the op itself (i.e., by canonicalization or folding)?

**Comment by hanhanW** [comment]
> I am not sure about this change. If outer dims perm is not a permutation in the first place, would it make more sense to drop it from the op itself (i.e., by canonicalization or folding)?

Good question. I had a similar question, and I decided to add the logic to the patterns. Here is what I'm thinking. We don't force people to run canonicalization patterns or folding patterns after every passes (or using createOrFold method). A pack/unpack op with identity outer dimension permutation is a valid op. The check here is quite simple, so I think the tradeoff (about adding the logic) is acceptable.

File: mlir/lib/Dialect/Utils/IndexingUtils.cpp [hunk 1]
```diff
@@ -213,6 +213,14 @@ mlir::invertPermutationVector(ArrayRef<int64_t> permutation) {
   return inversion;
 }
 
+bool mlir::isIdentityPermutation(ArrayRef<int64_t> permutation) {
+  int n = permutation.size();
```
Comment (chelini) [inline mlir/lib/Dialect/Utils/IndexingUtils.cpp:217]: Let's use int64_t to keep the type consistent with the element types in permutation.

**Review by chelini — APPROVED** [review]
Ok, thanks; it is a good tradeoff for now. I stumbled on this pattern as well:
```
%5 = tensor.empty() :  tensor<2x32xf32>
%pack_3 = tensor.pack %arg3 outer_dims_perm = [0] inner_dims_pos = [0] inner_tiles = [32] into %5 : tensor<64xf32> -> tensor<2x32xf32>
```
Would you mind adding it to the tests as well? It should fold too.

File: mlir/lib/Dialect/Utils/IndexingUtils.cpp (hunk 1 above)
Comment (hanhanW) [reply mlir/lib/Dialect/Utils/IndexingUtils.cpp:217]: I'm changing this to `for (auto i : llvm::seq<int64_t>(0, permutation.size()))` so we won't need `n`. And `i` uses int64_t type.

Comment (chelini) [reply mlir/lib/Dialect/Utils/IndexingUtils.cpp:217]: sounds good to me, thanks!

**Pull request merged**